### PR TITLE
Review Getting started instructions

### PR DESCRIPTION
**Pull request by snicoll**
This is a review of the getting started instructions for Wavefront. 

This covers several areas:
* Recommend the use of the BOM so that dependencies required by this starter are managed. In particular, if the version of the SDK micrometer uses is slightly behind, importing the BOM makes sure that the version of the SDK we intend is used
* Recommend to use start.spring.io for new projects. This is by far the easiest to get started. start.spring.io has also a "Expore" feature where you can see the content of a project without downloading it. That way you can peek at the build file and figure out yourself what you need to add to your project
* Stop hardcoding a version for Sleuth. This is problematic as those things move out of our control and  we can't remember to update the page whenever a new version is available (`2.2.5.RELEASE` is out today). Also, if a user already uses Spring Cloud they already have the Spring Cloud BOM so instructing them to add the starter with a version is a bad idea
* Move the building section down. Previously, building the project was required as we didn't have any snapshot, it shouldn't be front and center anymore.

@shavidissa I'd need your advices in particular on how to showcase the "Explore" part. I don't think I've phrased that clearly enough.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+The easiest way to get started is to create a new project on [start.spring.io](https://start.spring.io).
+Make sure to select Spring Boot 2.3 or above, your favorite build system, and include at least the "Wavefront" entry.
+If you want to opt-in for tracing support, add the "[Spring Cloud Sleuth](https://spring.io/projects/spring-cloud-sleuth)" entry as well.
 
-  ```shell script
-  $ ./mvnw clean install
-  ```
-* Start a sample web app on `localhost:8080` to see the basic usage of the starter. Invoke the following command in the root directory:
+If you already have a Spring Boot application, you can also use [start.spring.io](https://start.spring.io) to explore a project from your browser.
+This allows you to determine the setup for the Spring Boot generation your existing project is using.
```
Comment (rkempf42) [inline README.md:32]: This allows you to determine the setup for the Spring Boot generation your existing project is using. 
RK>>A little hard to parse. How about: 
RK>> This allows you to determine Spring Boot generation setup that your existing project is using.

File: README.md [hunk 2]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+
+```
+dependencyManagement {
+  imports {
+    mavenBom "com.wavefront:wavefront-spring-boot-bom:2.0.0"
+  }
+}
+```
+
+> Note: The `2.0.0` version is the latest version of the starter. start.spring.io maintains at all times which version of this starter should be used with the Spring Boot generation you are using.
```
Comment (rkempf42) [inline README.md:60]: Recommending Bold (**) for all notes and code (backticks) for all filenames, but your choice. 

RK>>
**Note**: The `2.0.0` version is the latest version of the starter. `start.spring.io` maintains at all times which version of this starter should be used with the Spring Boot generation you are using.

RK>>Don't completely understand this. What do we mean by "maintains"? Maybe others won't have a problem, but I'm finding it hard to grasp.

File: README.md [hunk 3]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+
+Connect to your Wavefront dashboard using this one-time use link:
+https://wavefront.surf/us/example
+```
+
+## Tracing Support
+
+If you'd like to send traces to Wavefront, you can do so using [Spring Cloud Sleuth](https://spring.io/projects/spring-cloud-sleuth) or [OpenTracing](https://opentracing.io/).
+
+To configure Spring Cloud in your existing project, you can either explore a new project on [start.spring.io](https://start.spring.io) and review the build configuration of your favorite build system.
```
Comment (rkempf42) [inline README.md:101]: To configure Spring Cloud in your existing project, you can either explore a new project on [start.spring.io](https://start.spring.io) and review the build configuration of your favorite build system.

RK>>Don't understand. If we say "either" shouldn't we say "or"? And, why "explore a new project"?

File: README.md [hunk 4]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+Each Spring Boot generation has a matching Spring Cloud generation, check [the project page](https://spring.io/projects/spring-cloud) for more details.
```
Comment (rkempf42) [inline README.md:102]: Each Spring Boot generation has a matching Spring Cloud generation, check [the project page](https://spring.io/projects/spring-cloud) for more details.

RK>> 2 sentences. 
Each Spring Boot generation has a matching Spring Cloud generation. Check [the project page](https://spring.io/projects/spring-cloud) for details.

File: README.md [hunk 5]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+
+Once you've added the `spring-cloud-dependencies` BOM, you can add Spring Cloud Sleuth as follows:
```
Comment (rkempf42) [inline README.md:104]: Once you've added
RK>>After you've added
RK>>We try to avoid terms that mean 2 things, such as "once" or "since" and use "after" and "because" instead in tech docs. You choice whether you want to do that.

File: README.md (hunk 1 above)
Comment (snicoll) [reply README.md:32]: I agree it's hard to parse but the suggestion doesn't convey what I meant, I think. It's not the "Spring Boot generation setup" but rather what you should be adding based on the Spring Boot generation you're using.

File: README.md (hunk 2 above)
Comment (snicoll) [reply README.md:60]: Yeah I re-read the sentence and it's hard to grasp. It's exactly the same thing as your first comment, actually. 

This is the core of this PR. Rather than hardcoding versions for stuff here, we need to ask the user to figure this out by themselves. We don't know which Spring Boot  generation they're using when they land on this `README` and we can't reasonably tell them about versions mapping in other projects that can easily be outdated.

Start.spring.io "maintains" this mapping. So if you go there, you select Spring Boot 2.4.x then add Wavefront, you can explore the project, peek at the build file and there you can see you need to use "X.Y.Z" (currently `2.0.0` but that could easily change shortly based on what we do). 

So that note is essentially saying "2.0.0 is the latest but it might not be what you should be using based on the Spring Boot generation you're using. You can look at what we do for a new project to figure out what to add in your own."

Better? If so, I'd love some insights on how to phrase that.

File: README.md (hunk 3 above)
Comment (snicoll) [reply README.md:101]: Yeah I guess I started a sentence with something in mind and half-finished it. The second sentence is the alternative, actually.

> why "explore a new project"?

You've already reviewed something where that's mentioned. Was it clear there and not here?

**Comment by snicoll** [comment]
@rkempf42 thanks for the review, I think we need one more round to phrase this properly.

File: README.md (hunk 4 above)
Comment (snicoll) [reply README.md]: Ack

File: README.md (hunk 5 above)
Comment (snicoll) [reply README.md]: Ack

File: README.md [hunk 6]
```diff
@@ -18,123 +19,141 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
-      ...
-      implementation 'io.opentracing.contrib:opentracing-spring-cloud-starter:0.5.3'
-
-    }
-    ```
+  </dependencies>
+</dependencyManagement>
+```
+
+If you are using Gradle, make sure you project uses the `io.spring.dependency-management` plugin and add the following to your `build.gradle` file:
```
Comment (rkempf42) [inline README.md:50]: ...make sure your project...

File: README.md (hunk 1 above)
Comment (rkempf42) [reply README.md:32]: Maybe: You can see the current Spring Boot generation setup and explore dependencies.

File: README.md (hunk 2 above)
Comment (rkempf42) [reply README.md]: Yes, your suggestion is very close. How about this: 
2.0.0 is the latest version. However, the actual version you need depends on the Spring Boot generation you're using. Spring Initializr helps you explore what makes sense for your environment.

File: README.md (hunk 3 above)
Comment (rkempf42) [reply README.md]: Yes, I understand it now. Thanks!

**Comment by snicoll** [comment]
@shavidissa I've pushed an update. Could you please let me know what you think?

File: README.md [hunk 7]
```diff
@@ -16,125 +17,147 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+If you already have a Spring Boot application, you can also use [start.spring.io](https://start.spring.io) to explore a new project from your browser.
+This allows you to see the setup for the Spring Boot generation your project is using. 
 
-  ```shell script
-  $ ./mvnw spring-boot:run -pl wavefront-spring-boot-sample
-  ```
+For completeness, here is what you should follow to configure your project.
 
-* If you already have a Spring Boot application, be sure to use Spring Boot 2.3 or later and add the following dependency to your `pom.xml` file:
+> **Note**: The `2.0.1` version is the latest version of the starter. [start.spring.io](https://start.spring.io) maintains at all times which version of this starter should be used with the Spring Boot generation you are using.
```
Comment (shavidissa) [inline README.md:38]: How about the following? I know its a bit too long but wanted to be a bit more clear just in case. If too long, we can remove the last sentence.

> **Note**: The 2.0.1 version is the latest version of the starter. The Wavefront for Spring Boot dependency needs to be compatible with the Spring Boot release version. [start.spring.io](https://start.spring.io) maintains the up to date release version.
>
> To make sure that you are using the correct version, go to start.spring.io and create a new project with the Wavefront dependency and explore the `pom.xml` or `build.gradle` file.

File: README.md [hunk 8]
```diff
@@ -16,125 +17,147 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
 
-    ```xml
+The core setup consists of importing the `wavefront-spring-boot` Bill Of Materials (BOM).
```
Comment (shavidissa) [inline README.md:40]: Suggestion to format it as follows:
-------------------

Configure your project using Maven or Gradle.

**Maven**

 - The core setup consists of importing the `wavefront-spring-boot` Bill Of Materials (BOM).

  ```<dependencyManagement>
    <dependencies>
      <dependency>
        <groupId>com.wavefront</groupId>
        <artifactId>wavefront-spring-boot</artifactId>
        <version>2.0.1</version>
        <type>pom</type>
        <scope>import</scope>
      </dependency>
    </dependencies>
  </dependencyManagement>
  ```

- Add the `wavefront-spring-boot-starter` to your project.

  ```
  <dependency>
    <groupId>com.wavefront</groupId>
    <artifactId>wavefront-spring-boot-starter</artifactId>
  </dependency>
  ```

**Gradle**

- If you are using Gradle, make sure your project uses the `io.spring.dependency-management` plugin and add the following to your build.gradle file:

  ```
  dependencyManagement {
    imports {
      mavenBom "com.wavefront:wavefront-spring-boot-bom:2.0.1"
    }
  }
  ```
- Add the `wavefront-spring-boot-starter` to your project.

  ```
  dependencies {
    ...
    implementation 'com.wavefront:wavefront-spring-boot-starter'

  }
  ```

File: README.md [hunk 9]
```diff
@@ -16,125 +17,147 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+```
+
+## Tracing Support
+
+If you'd like to send traces to Wavefront, you can do so using [Spring Cloud Sleuth](https://spring.io/projects/spring-cloud-sleuth) or [OpenTracing](https://opentracing.io/).
+
+To configure Spring Cloud in your existing project, you can explore a new project on [start.spring.io](https://start.spring.io) and review the build configuration for your favorite build system.
+Each Spring Boot generation has a matching Spring Cloud generation. Check [the project page](https://spring.io/projects/spring-cloud#release-trains) for more details.
+
+After you've added the `spring-cloud-dependencies` BOM, you can add Spring Cloud Sleuth as follows:
```
Comment (shavidissa) [inline README.md:108]: Formatting Suggestion
-----------------------------------------
**Spring Cloud Sleuth**

Each Spring Boot generation has a matching Spring Cloud generation. You can run a simple command and find out the Spring Cloud version that is compatible with your Spring Boot version. See [Spring Cloud Version Given Spring Boot Version](https://spring-cloud-info.cfapps.io/docs/spring-cloud-info.html#_spring_cloud_version_given_spring_boot_version) for details. 

After you've added the spring-cloud-dependencies BOM, you can add Spring Cloud Sleuth as follows:

- Maven: Add the following dependency to the `pom.xml` file
  ```
  <dependency>
    <groupId>org.springframework.cloud</groupId>
    <artifactId>spring-cloud-starter-sleuth</artifactId>
  </dependency>
  ```
  
- Gradle: Add the following dependency to the `build.gradle` file:

  ```
  dependencies {
    ...
    implementation 'org.springframework.cloud:spring-cloud-starter-sleuth'
  }
  ```

**OpenTracing**  
Configure your `pom.xml` file or the `build.gradle` file.

- Maven: Add the following dependencies to your `pom.xml` file:

  ```
  <dependency>
    <groupId>io.opentracing.contrib</groupId>
    <artifactId>opentracing-spring-cloud-starter</artifactId>
    <version>0.5.7</version>
  </dependency>
  ```

- Gradle: Add the following dependencies to the `build.gradle` file:

  ```
  dependencies {
    ...
    implementation 'io.opentracing.contrib:opentracing-spring-cloud-starter:0.5.7'
  }
  ```

File: README.md [hunk 10]
```diff
@@ -16,125 +17,147 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+
+Connect to your Wavefront dashboard using this one-time use link:
+https://wavefront.surf/us/example
+```
+
+## Tracing Support
+
+If you'd like to send traces to Wavefront, you can do so using [Spring Cloud Sleuth](https://spring.io/projects/spring-cloud-sleuth) or [OpenTracing](https://opentracing.io/).
+
+To configure Spring Cloud in your existing project, you can explore a new project on [start.spring.io](https://start.spring.io) and review the build configuration for your favorite build system.
```
Comment (shavidissa) [inline README.md:105]: Update it based on the info shared by Ryan:

Each Spring Boot generation has a matching Spring Cloud generation. You can run a simple command and find out the Spring Cloud version that is compatible with your Spring Boot version. See [Spring Cloud Version Given Spring Boot Version](https://spring-cloud-info.cfapps.io/docs/spring-cloud-info.html#_spring_cloud_version_given_spring_boot_version) for details.

File: README.md [hunk 11]
```diff
@@ -16,125 +17,147 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
+```
+dependencies {
+  ...
+  implementation 'io.opentracing.contrib:opentracing-spring-cloud-starter:0.5.7'
+
+}
+```
+
+## Building
+To build the latest state of this project, invoke the following command from the root directory:
```
Comment (shavidissa) [inline README.md:148]: Suggestion: Shall we add the steps for Gradle too? :)

Comment (snicoll) [reply README.md:148]: This project does not use Gradle, only Maven.

**Comment by snicoll** [comment]
@shavidissa thanks a lot for the review. I've pushed one more commit. It has the link to the production doc that might not be live yet. I hope the formatting is correct now? Let me know what you think.

Thanks!

**Comment by shavidissa** [comment]
Hi @snicoll,
This looks good. :) We can release or Wavefront docs too when this is going live so the link won't break here.

File: README.md [hunk 12]
```diff
@@ -16,125 +17,160 @@ This project provides a Spring Boot starter for Wavefront. Add the starter to a
...
-    dependencies {
-      ...
-      implementation 'io.opentracing.contrib:opentracing-spring-cloud-starter:0.5.3'
+If you'd like to send traces to Wavefront, you can do so using [Spring Cloud Sleuth](https://spring.io/projects/spring-cloud-sleuth) or [OpenTracing](https://opentracing.io/).
 
-    }
-    ```
+**Spring Cloud Sleuth**
+
+Each Spring Boot generation has a matching Spring Cloud generation.
```
Comment (shavidissa) [inline README.md:113]: Since we have a good doc link now:

Each Spring Boot generation has a matching Spring Cloud generation. See <a href="https://spring.io/projects/spring-cloud#getting-started">Getting Started on the Spring Cloud documentation</a> for details.
